src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVMKernels.cpp

```diff
@@ -663,6 +663,356 @@ INST_FluxKernel( 5, FaceElementStencil );
+                   arraySlice2d< real64 const > dPhaseRelPerm_dPhaseVolFrac,
+                   arraySlice1d< real64 const > phaseVisc,
+                   arraySlice1d< real64 const > phaseOutflux,
+                   real64 & phaseCFLNumber )
+{
+  // first, check which phases are mobile in the cell
+  real64 mob[NP]{};
+  localIndex mobilePhases[NP]{};
+  localIndex numMobilePhases = 0;
+  for( localIndex ip = 0; ip < NP; ++ip )
+  {
+    mob[ip] = phaseRelPerm[ip] / phaseVisc[ip];
+    if( mob[ip] > 1e-12 )
+    {
+      mobilePhases[numMobilePhases] = ip;
+      numMobilePhases++;
+    }
+  }
+
+  // then, depending on the regime, apply the appropriate CFL formula
+  phaseCFLNumber = 0;
+
+  // single-phase flow regime
+  if( numMobilePhases == 1 )
+  {
+    phaseCFLNumber = phaseOutflux[mobilePhases[0]] / poreVol;
+  }
+  // two-phase flow regime
+  else if( numMobilePhases == 2 )
+  {
+    // from Hui Cao's PhD thesis
+    localIndex const ip0 = mobilePhases[0];
+    localIndex const ip1 = mobilePhases[1];
+    real64 const dMob_dVolFrac[2] = { dPhaseRelPerm_dPhaseVolFrac[ip0][ip0] / phaseVisc[ip0],
+                                      -dPhaseRelPerm_dPhaseVolFrac[ip1][ip1] / phaseVisc[ip1] }; // using S0 = 1 - S1
+    real64 const denom = 1. / ( poreVol * ( mob[ip0] + mob[ip1] ) );
+    real64 const coef0 = denom * mob[ip1] / mob[ip0] * dMob_dVolFrac[ip0];
+    real64 const coef1 = -denom * mob[ip0] / mob[ip1] * dMob_dVolFrac[ip1];
+
+    phaseCFLNumber = fabs( coef0*phaseOutflux[ip0] + coef1*phaseOutflux[ip1] );
+  }
+  // three-phase flow regime
+  else if( numMobilePhases == 3 )
+  {
+    // from Keith Coats, IMPES stability: Selection of stable timesteps (2003)
+    real64 totalMob = 0.0;
+    for( localIndex ip = 0; ip < numMobilePhases; ++ip )
+    {
+      totalMob += mob[ip];
+    }
+
+    real64 f[2][2]{};
+    for( localIndex i = 0; i < 2; ++i )
+    {
+      for( localIndex j = 0; j < 2; ++j )
+      {
+        f[i][j]  = ( i == j )*totalMob - mob[i];
+        f[i][j] /= (totalMob * mob[j]);
+        real64 sum = 0;
+        for( localIndex k = 0; k < 3; ++k )
+        {
+          sum += dPhaseRelPerm_dPhaseVolFrac[k][j] / phaseVisc[k]
+                 * phaseOutflux[j];
+        }
+        f[i][j] *= sum;
+      }
+    }
+    phaseCFLNumber = f[0][0] + f[1][1];
+    phaseCFLNumber += sqrt( phaseCFLNumber*phaseCFLNumber - 4 * ( f[0][0]*f[1][1] - f[1][0]*f[0][1] ) );
+    phaseCFLNumber = 0.5 * fabs( phaseCFLNumber ) / poreVol;
+  }
+}
+
+
+template< localIndex NC >
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void
+CFLKernel::
+  computeCompCFL( real64 const & poreVol,
+                  arraySlice1d< real64 const > compDens,
+                  arraySlice1d< real64 const > compOutflux,
+                  real64 & compCFLNumber )
+{
+
+
+  compCFLNumber = 0.0;
+  for( localIndex ic = 0; ic < NC; ++ic )
+  {
+    real64 const compMoles = compDens[ic] * poreVol;
+    if( compMoles > 1e-12 )
```

Comment:
Same here I suppose. Also this threshold seems rather scale-dependent, I guess we could use component fraction as an "intrinsic" quantity to detect present components, which doesn't protect against near-zero pore volume, but that would be a problem for other reasons...

Reply:
Now I check the component fraction and not anymore the moles of the components in this threshold, and I added a class-level constant as you suggest.